src/ray/raylet/node_manager.cc

```diff
@@ -1693,35 +1757,88 @@ void NodeManager::FinishAssignedActorTask(Worker &worker, const Task &task) {
+    if (!object_to_release.is_nil()) {
+      // If there were no new actor handles created, then no other actor task
+      // will depend on this execution dependency, so it safe to release.
+      HandleObjectMissing(object_to_release);
+    }
+    // Mark the dummy object as locally available to indicate that the actor's
+    // state has changed and the next method can run. This is not added to the
+    // object table, so the update will be invisible to both the local object
+    // manager and the other nodes.
+    // NOTE(swang): The dummy objects must be marked as local whenever
+    // ExtendFrontier is called, and vice versa, so that we can clean up the
+    // dummy objects properly in case the actor fails and needs to be
+    // reconstructed.
+    HandleObjectLocal(dummy_object);
+  } else {
+    // If this actor was resumed from a checkpoint.
+    // Look up the checkpoint in GCS and use it to restore actor registration.
+    const auto checkpoint_id = checkpoint_id_to_restore_[actor_id];
+    checkpoint_id_to_restore_.erase(actor_id);
+    RAY_LOG(DEBUG) << "Looking up checkpoint " << checkpoint_id << " for actor "
+                   << actor_id;
+    RAY_CHECK_OK(gcs_client_->actor_checkpoint_table().Lookup(
+        JobID::nil(), checkpoint_id,
+        [this, actor_entry](ray::gcs::AsyncGcsClient *client,
+                            const UniqueID &checkpoint_id,
+                            const ActorCheckpointDataT &checkpoint_data) {
+          RAY_LOG(INFO) << "Restoring registration for actor " << actor_entry->first
+                        << " from checkpoint " << checkpoint_id;
+          actor_entry->second.RestoreFrontier(checkpoint_data);
+          // Mark the unreleased dummy objects as local.
+          for (const auto &entry : actor_entry->second.GetDummyObjects()) {
+            HandleObjectLocal(entry.first);
+          }
+          // The actor's state is now restored. Resubmit the waiting actor tasks, because
+          // some tasks may be before the checkpoint, and will otherwise get stuck in the
+          // waiting queue forever.
+          std::unordered_set<TaskID> waiting_actor_task_ids;
+          for (const auto &task : local_queues_.GetTasks(TaskState::WAITING)) {
+            if (task.GetTaskSpecification().ActorId() == actor_entry->first) {
+              waiting_actor_task_ids.insert(task.GetTaskSpecification().TaskId());
+            }
+          }
+          for (const auto &task : local_queues_.RemoveTasks(waiting_actor_task_ids)) {
+            if (!lineage_cache_.RemoveWaitingTask(
+                    task.GetTaskSpecification().TaskId())) {
+              RAY_LOG(WARNING) << "Task " << task.GetTaskSpecification().TaskId()
+                               << " already removed from the lineage cache. This is most "
+                                  "likely due to reconstruction.";
+            }
+            SubmitTask(task, Lineage());
+          }
```

Comment:
@stephanie-wang I ended up fixing the issue by resubmitting the waiting task here.

Reply:
Hmm okay, I think I prefer the other solution I mentioned because it seems cleaner, but maybe I'm wrong. I'll give it a shot and push to the PR if it seems doable.